_posts/2017-10-04-Law-of-quality.md

```diff
@@ -0,0 +1,106 @@
+---
+layout: post
+title: The Law of Three for Software Quality
+author: lokori
+excerpt: There is a simple rule-of-thumb for software quality which seems to remain true over the years: 
+If the system properly handles dates, characters and money, it has good quality. 
+Short of scientific evidence, empirical arguments are provided. Try and see how your system measures up against this law.
+tags:
+- quality
+- bugs
+- encoding
+- SQL
+- Java
+---
+
+## The Law of Three
+
+This is the law. 
+
+### If the system properly handles dates, characters and money, it has good quality.
+
+I have been talking about it every now and then for quite some time. It seems to hold true in the world of professional software development where I work as years pass by. Most of the real world systems handle characters and dates in some fashion, and getting these two right is a major challenge even if money is not involved. The law sets a sufficient condition; I rarely see a software system of non-trivial size which got all these right. 
+
+This post examines the difficulties in these three matters to shed some light into why it's actually quite difficult. The intention is not deal with these subjects a in thorough manner - that would fill a book I'm not qualified to write.
+
+## Dates
+
+Date and time are something familiar to us, but the devil is in the details. [The Long, Painful History of Time](http://naggum.no/lugm-time.html) gives a good account of the intricacies of date and time.
+
+Provided you understand this historical context and time zones, this is only half of the victory. Let's consider Java and a standard JDBC interface to a relation database, say PostgreSQL. 
+
+Originally Java had two Date classes: ```java.util.Date``` and ```java.sql.Date```. Despite Java core developers being smart, it
+turned out that poor ```java.util.Date``` had issues. A decade later Java time API was rewritten in [JSR 310](https://community.oracle.com/docs/DOC-983209). And then [JodaTime](http://www.joda.org/joda-time/) came to rescue the suffering Java community. Now the confused developer has deprecated (and broken) legacy classes laying around, some newer ones to choose from and a lot of headache. 
+
+Okay, but surely the situation is better in SQL databases? We'll define our table like this:
+```
+CREATE TABLE dadas
+  (
+    le_time TIMESTAMPT NOT NULL
+  ) ;
+```
+
+Oops, we just failed. According to [PostgreSQL documentation](https://www.postgresql.org/docs/9.1/static/datatype-datetime.html) this leaves out timezone, which means effectively system locale. 
```

Comment:
“which means…” -> ”which effectively means the system locale”

Reply:
aye